File: app/src/main/java/net/squanchy/eventdetails/EventDetailsService.java
package net.squanchy.eventdetails;

import com.google.firebase.auth.FirebaseUser;

import net.squanchy.schedule.domain.view.Event;
import net.squanchy.service.firebase.FirebaseAuthService;
import net.squanchy.service.repository.EventRepository;
import net.squanchy.support.lang.Optional;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

class EventDetailsService {

    private final EventRepository eventRepository;
    private final FirebaseAuthService authService;

    EventDetailsService(EventRepository eventRepository, FirebaseAuthService authService) {
        this.eventRepository = eventRepository;
        this.authService = authService;
    }

    public Observable<Event> event(String eventId) {
        return authService.ifUserSignedInThenObservableFrom(userId -> eventRepository.event(eventId, userId));
    }

    Single<FavoriteResult> toggleFavorite(Event event) {
        return currentUser()
                .flatMap(optionalUser -> {
                    if (!optionalUser.isPresent() || optionalUser.get().isAnonymous()) {
                        return Single.just(FavoriteResult.MUST_AUTHENTICATE);
                    }

                    return toggleFavoriteOn(event)
                            .andThen(Single.just(FavoriteResult.SUCCESS));
                });
    }

    private Completable toggleFavoriteOn(Event event) {
        if (event.favorited()) {
            return removeFavorite(event.id());
        } else {
            return favorite(event.id());
        }
    }

    private Completable removeFavorite(String eventId) {
        return authService.ifUserSignedInThenCompletableFrom(userId -> eventRepository.removeFavorite(eventId, userId));
    }

    private Completable favorite(String eventId) {
        return authService.ifUserSignedInThenCompletableFrom(userId -> eventRepository.addFavorite(eventId, userId));
    }

    private Single<Optional<FirebaseUser>> currentUser() {
        return authService.currentUser().firstOrError();
    }

    enum FavoriteResult {
        SUCCESS,
        MUST_AUTHENTICATE
    }
}

File: app/src/main/java/net/squanchy/eventdetails/EventDetailsService.kt
package net.squanchy.eventdetails

import android.annotation.SuppressLint
import com.google.firebase.auth.FirebaseUser
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import net.squanchy.schedule.domain.view.Event
import net.squanchy.service.firebase.FirebaseAuthService
import net.squanchy.service.repository.EventRepository
import net.squanchy.support.lang.Optional

internal class EventDetailsService(
        private val eventRepository: EventRepository,
        private val authService: FirebaseAuthService
) {

    fun event(eventId: String): Observable<Event> {
        return authService.ifUserSignedInThenObservableFrom { userId -> eventRepository.event(eventId, userId) }
    }

    @SuppressLint("CheckResult") // False positive, to remove in 3.1.0-beta5
    fun toggleFavorite(event: Event): Single<FavoriteResult> {
        return currentUser()
            .flatMap { optionalUser ->
                optionalUser
                    .map {
                        if (it.isAnonymous) {
                            Single.just(FavoriteResult.MUST_AUTHENTICATE)
                        } else {
                            toggleFavoriteOn(event).andThen(Single.just(FavoriteResult.SUCCESS))
                        }
                    }
                    .or(Single.just(FavoriteResult.MUST_AUTHENTICATE))
            }
    }

    private fun toggleFavoriteOn(event: Event): Completable {
        return if (event.favorited) {
            removeFavorite(event.id)
        } else {
            favorite(event.id)
        }
    }

    private fun removeFavorite(eventId: String): Completable {
        return authService.ifUserSignedInThenCompletableFrom { userId -> eventRepository.removeFavorite(eventId, userId) }
    }

    private fun favorite(eventId: String): Completable {
        return authService.ifUserSignedInThenCompletableFrom { userId -> eventRepository.addFavorite(eventId, userId) }
    }

    private fun currentUser(): Single<Optional<FirebaseUser>> = authService.currentUser().firstOrError()

    internal enum class FavoriteResult {
        SUCCESS,
        MUST_AUTHENTICATE
    }
}
